File: worker/src/srs.ts
import { todayStr, addDays } from './db'
import type { Env } from './router'

const RATINGS: Record<string, number> = { esqueci: 0, dificil: 1, facil: 2, muito_facil: 3 }

export async function applyReview(env: Env, card: any, rating: string): Promise<Record<string, unknown>> {
  const g = RATINGS[rating] ?? 2
  let reps = card.reps + 1
  let lapses = card.lapses
  let ease = card.ease
  let interval = card.interval_days
  let due: string
  if (g === 0) {
    lapses += 1
    ease = Math.max(1.3, ease - 0.2)
    interval = 0
    reps = 0
    due = todayStr()
  } else {
    if (reps === 1) interval = 1
    else if (reps === 2) interval = 3
    else if (g === 1) {
      interval = Math.max(1, Math.round(interval * 1.2))
      ease = Math.max(1.3, ease - 0.15)
    } else if (g === 2) {
      interval = Math.max(1, Math.round(interval * ease))
    } else {
      interval = Math.max(1, Math.round(interval * ease * 1.3))
      ease = Math.min(3.0, ease + 0.15)
    }
    due = addDays(todayStr(), interval)
  }
  await env.DB.prepare('UPDATE flashcards SET reps = ?, lapses = ?, ease = ?, interval_days = ?, due = ? WHERE id = ?')
    .bind(reps, lapses, ease, interval, due, card.id).run()
  return { interval_days: interval, ease, reps, lapses, due }
}

File: backend/app/services/srs.py
"""Sistema de repetição espaçada (SM-2 adaptado, estilo Anki)."""
from datetime import date, timedelta
from sqlalchemy.orm import Session
from ..models import Flashcard

RATINGS = {
    "esqueci": 0,
    "dificil": 1,
    "facil": 2,
    "muito_facil": 3,
}


def apply_review(card: Flashcard, rating: str) -> dict:
    g = RATINGS.get(rating, 2)
    card.reps += 1
    if g == 0:
        card.lapses += 1
        card.ease = max(1.3, card.ease - 0.2)
        card.interval_days = 0
        card.reps = 0
        card.due = date.today()
    else:
        if card.reps == 1:
            card.interval_days = 1
        elif card.reps == 2:
            card.interval_days = 3
        else:
            if g == 1:
                card.interval_days = max(1, round(card.interval_days * 1.2))
                card.ease = max(1.3, card.ease - 0.15)
            elif g == 2:
                card.interval_days = max(1, round(card.interval_days * card.ease))
            else:
                card.interval_days = max(1, round(card.interval_days * card.ease * 1.3))
                card.ease = min(3.0, card.ease + 0.15)
        card.due = date.today() + timedelta(days=card.interval_days)
    return {
        "interval_days": card.interval_days,
        "ease": card.ease,
        "reps": card.reps,
        "lapses": card.lapses,
        "due": card.due.isoformat(),
    }
